[addons] fix multiple usage

AlwinEsch opened the pull request:
## Description
If one addon defines several types inside, was before always only the first
possible to start, e.g. screensaver.atv4 has also a video addon inside,
but this was before not possible to start.

The view, control and start of addons where have more as one type inside are now fixed with this request.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):
Now with more correct view:
![Bildschirmfoto von 2020-01-28 01-50-32](https://user-images.githubusercontent.com/6879739/73226743-6b910f00-4171-11ea-8cd9-1189ce09a1f5.png)



## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

enen92 commented on xbmc/addons/PluginSource.cpp:
```diff
@@ -102,7 +102,8 @@ bool CPluginSource::IsType(TYPE type) const
        || (type == ADDON_GAME && Provides(GAME))
-       || (type == ADDON_EXECUTABLE && Provides(EXECUTABLE)));
+       || (type == ADDON_EXECUTABLE && Provides(EXECUTABLE))
+       || CAddon::IsType(type)); // Fallback also to other types e.g. ADDON_SCRIPT is asked
```
Why is this change needed?

enen92 replied:
It doesn't make much sense to me. A plugin is a plugin, why do you check it against `ADDON_SCRIPT` ?

AlwinEsch replied:
The directory ask always for `ADDON_SCRIPT` but the scripts are also handled by `CPluginSource` where primary for `ADDON_PLUGIN`, without it becomes never a true if asked for.

enen92 commented on xbmc/addons/Addon.h:
```diff
@@ -35,7 +35,7 @@ class CAddon : public IAddon
   TYPE Type() const override { return m_addonInfo->MainType(); }
   TYPE FullType() const override { return Type(); }
-  bool IsType(TYPE type) const override { return type == m_addonInfo->MainType(); }
+  bool IsType(TYPE type) const override { return m_addonInfo->IsType(type); }
```
Wouldn't it be easier to:

- Change `Type()` function to `MainType()` as the function name is misleading
- Create a new method `bool HasType(Type)` that would return `true` if an addon has the given type
- Just use HasType(xx) in the switch/if conditions

?

AlwinEsch replied:
With `Type()`I can change, was only leaved as them to have the cpluff remove request not bigger with rename of them.

With new method `bool HasType(Type)`, what should be different to `bool IsType(TYPE type)`? 

Or you mean to leave the ` { return type == m_addonInfo->MainType(); }`, this was I think only added from me to prevent this view the 2 new types on addon Window (If I remember correctly)?

enen92 replied:
I just gave it a try and it does indeed fix the issue with `screensaver.atv4`. However I don't like the Scripts and Media sources showing in the addon browser.

What I mean with my suggestion was exactly that. `Type()` and `IsType()` refer to MainType only so they should be renamed to reflect what they do. `HasType` (or your new `IsType`) would have the implementation you provide in this PR. The only code block using `HasType` would be `GetScriptsAndPlugins` for now.

enen92 commented:
JFY: Issue with `embruary.info` still persists after this PR:

![alt text](https://i.imgur.com/qXWlwdP.png "Logo Title Text 1")

If you install the addon from the official repo, you'll find 2 icons under video addons. One of the entrypoints is a script (and should show under Program Addons)

enen92 commented:
Fixes https://github.com/xbmc/xbmc/issues/17125 (cross-reference)

@AlwinEsch I know you're quite busy with binary addons. I'm wondering if you could pick this up whenever possible and find a way of not presenting "Scripts" and "Media Sources" in the interface. Unfortunately without this getting fix I can't work on my screensaver.
Cheers

AlwinEsch commented:
I hope to go soon back on them.

AlwinEsch committed:
[addons] prevent on addon directory the add of "plugin" or "script" as own folder

This types have always a subtype where controlled in own folders.
Needed for following changes.

AlwinEsch committed:
[addons] remove func. "FullType()" from IAddon and other places, becomes nowhere used

AlwinEsch committed:
[addons] rename all addon IsType func. to HasType (is more correct)

garbear commented on xbmc/addons/PluginSource.cpp:
```diff
@@ -80,29 +86,11 @@ CPluginSource::Content CPluginSource::Translate(const std::string &content)
-       || (type == ADDON_EXECUTABLE && Provides(EXECUTABLE)));
+  return ((type == ADDON_VIDEO && Provides(VIDEO)) || (type == ADDON_AUDIO && Provides(AUDIO)) ||
+          (type == ADDON_IMAGE && Provides(IMAGE)) || (type == ADDON_GAME && Provides(GAME)) ||
+          (type == ADDON_EXECUTABLE && Provides(EXECUTABLE)) || (type == CAddon::Type()));
```
Should we keep these on separate lines for readability and git history niceness?

AlwinEsch replied:
Aha github is working again :smile: 
No problem hold it also for better like before was only about the guidelines. 
I change it.

AlwinEsch commented:
This request is complete udpated and have mostly now the same view as on Leia, also was the test with Video and App addons OK.

Also renamed the IsType to HasType, does more match the wanted work.

Here is also (for testing) this request changes on Version with Python 2: https://github.com/AlwinEsch/kodi/tree/fix-multi-addons-python2

AlwinEsch committed:
[addons] fix script/plugin subtype check

Before was wrongly during creation the subtype added to addon info,
this then make e.g. "video" to all types inside a addon and "program" then
also declared wrongly as "video".

AlwinEsch committed:
[addons] fix check on CAddonsDirectory::GetScriptsAndPlugins for right type

This fix the ask about the script and plugin addons to have the right
start path, that they can be used on addon window.

This failure effect was present in case a addon has contained more as one
type, e.g. python screensaver together with video addon.

AlwinEsch committed:
[addons] use in filesystem the IsAddonInstalled to check

No need to call "CServiceBroker::GetAddonMgr().GetAddon(addon->ID(), localAddon, ADDON_UNKNOWN, false)"
if already a function in Manager where do this.

AlwinEsch committed:
[addons] add a bit of documentation to CAddon and CAddonInfo

This to have a description what the related functions do.

enen92 reviewed (approved):
I can confirm this fixes https://github.com/xbmc/xbmc/issues/17125 and also the issue affecting embruary.info. Thanks a lot!
